### PR TITLE
Recurse submodules on branch checkout

### LINKED ISSUE #4579: Switching branches does not update submodules
## Description
When switching between branches of a repo, submodules are not updated

## Version
* GitHub Desktop: 1.1.1
* Operating system: macOS 10.13.4

## Steps to Reproduce

1. In a repo called `repo`, with branches `foo` and `bar`, add a submodule `sub`, with branches `subfoo` and `subbar`.
2. In branch `bar` checkout `subbar` in the submodule and commit
3. In branch `foo` checkout `subfoo` in the submodule and commit
4. In GitHub Desktop, switch between the `foo` and `bar` branches

### Expected Behavior
GitHub Desktop runs `git submodule update` on every branch switch, so that branch `foo` always uses branch `subfoo` of the submodule and branch `bar` always uses branch `subbar` of the submodule.

### Actual Behavior
The submodule is not updated during branch switches. GitHub Desktop will mark the submodule as changed.

### PR DESCRIPTION
**Pull request by jessetan**
This PR adds `--recurse-submodules` to the options of `git checkout`, `pull` and `fetch`, so that submodules are correctly updated when switching branches.
This makes submodules "just work" a little better. Without this, the user needs to update submodules manually, which is not possible using GitHub Desktop.

Or as written on the Git mailing list:
> This is a must if you are working with submodules, otherwise every git
checkout requires a git submodule update, and you forget it, and
things break, and you understand, and you go to stack overflow
questions 
http://stackoverflow.com/questions/22328053/why-doesnt-git-checkout-automatically-do-git-submodule-update-recursive
http://stackoverflow.com/questions/4611512/is-there-a-way-to-make-git-pull-automatically-update-submodules
and you give up and create aliases :-)

Fixes #4579

### CODE REVIEW

**Comment by shiftkey** [comment]
@jessetan just a heads up - those CI errors are because `prettier` enforces a max width when emitting, but you can quickly fix this by running `yarn lint:prettier --fix` and committing the change

**Comment by shiftkey** [comment]
@jessetan apologies for the delays in reviewing this.

I mostly wanted to dig into how Git fails when submodules are not in a known good state, because I wanted to avoid scenarios where users checkout a branch and this operation fails consistently because:

 - submodule remotes are no longer accessible
 - submodule remotes are not on GitHub and require basic authentication (does Desktop handle this?)
 - probably some others, but I need to rack my brain

Hopefully these are covered by exit codes or error messages that we can handle in the app and figure out whether to interrupt the user using our error-handling infrastructure.

**Comment by shiftkey** [comment]
I've bumped this PR with some tests for `checkout` and ensuring things are correct afterwards, but testing `fetch/pull` was tripping our `GIT_ASKPASS` handler (because things like missing repositories will instead try to authenticate again) so errors there will end up in our normal error handling.

Dropping myself from the review, but I think this is ready for a proper review now.

**Comment by shiftkey** [comment]
:apple:

File: app/src/lib/git/checkout.ts [hunk 1]
```diff
@@ -69,8 +69,14 @@ export async function checkoutBranch(
 
   const args =
     branch.type === BranchType.Remote
-      ? baseArgs.concat(branch.name, '-b', branch.nameWithoutRemote, '--')
-      : baseArgs.concat(branch.name, '--')
+      ? baseArgs.concat(
```
[the remark this answers is not in the document]
Comment (shiftkey) [reply app/src/lib/git/checkout.ts]: I don't believe so - the behaviour is driven by the feature flag, and its not really something that the user (or developer) can control currently.

**Comment by iAmWillShepherd** [comment]
Putting this in the next release so we can get some time in beta before moving to production.

**Pull request merged**